CodeGeneration/Sources/SyntaxSupport/CompatibilityLayers.swift

```diff
@@ -0,0 +1,192 @@
+          Child(newerChildPath: [newerChild, newerGrandchild])
+        }
+      }
+    }
+
+    if cachedReplacementChildren[newerChild] == nil {
+      cachedReplacementChildren[newerChild] = make()
+    }
+    return cachedReplacementChildren[newerChild]!
+  }
+  private var cachedReplacementChildren: [Child: [Child]] = [:]
+
+  /// Compute and cache compatibility layer information for the given node, unless it is already present.
+  private mutating func realizeLayers(for node: Node) {
+    guard deprecatedVarsByNode[node.syntaxNodeKind] == nil, let layoutNode = node.layoutNode else {
+      return
+    }
+
+    let result = computeLayersFor(
+      typeName: layoutNode.kind.rawValue,
+      initialChildren: layoutNode.children,
+      history: layoutNode.childHistory,
+      areRequirements: false
+    )
+
+    deprecatedVarsByNode[node.syntaxNodeKind] = result.vars
+    deprecatedInitSignaturesByNode[node.syntaxNodeKind] = result.initSignatures
+  }
+
+  private mutating func realizeLayers(for trait: Trait) {
+    guard deprecatedVarsByTrait[trait.traitName] == nil else {
+      return
+    }
+
+    let result = computeLayersFor(
+      typeName: trait.traitName,
+      initialChildren: trait.children,
+      history: trait.childHistory,
+      areRequirements: true
+    )
+
+    deprecatedVarsByTrait[trait.traitName] = result.vars
+    deprecatedInitSignaturesByTrait[trait.traitName] = result.initSignatures
+  }
+
+  /// Compute and cache compatibility layer information for the given children.
+  private mutating func computeLayersFor(
+    typeName: String,
+    initialChildren: [Child],
+    history: Child.History,
+    areRequirements: Bool
+  ) -> (vars: [Child], initSignatures: [InitSignature]) {
+    // The results that will ultimately be saved into the *ByNode dictionaries.
+    var vars: [Child] = []
+    var initSignatures: [InitSignature] = []
+
+    // Temporary working state for the loop.
+    var children = initialChildren
+    var knownVars = Set(children)
+
+    func firstIndexOfChild(named targetName: String) -> Int {
+      guard let i = children.firstIndex(where: { $0.name == targetName }) else {
+        fatalError("couldn't find '\(targetName)' in current children of \(typeName): \(String(reflecting: children.map(\.name)))")
+      }
+      return i
+    }
+
+    for changeSet in history {
+      var unexpectedChildrenWithNewNames: Set<Child> = []
+
+      // First pass: Apply the changes explicitly specified in the change set.
+      for (currentName, refactoring) in changeSet {
+        let i = firstIndexOfChild(named: currentName)
+
+        let replacementChildren = replacementChildren(for: children[i], by: refactoring)
+        children.replaceSubrange(i...i, with: replacementChildren)
+
+        if !areRequirements {
+          // Mark adjacent unexpected node children whose names have changed too.
+          if currentName != (replacementChildren.first?.name ?? "") {
```

Comment:
I don’t think you need the `?? ""` here. We should be able to compare the non-optional `currentName` to an optional on the rhs. (same a few lines below)